src/state/service.rs

```diff
@@ -188,59 +201,69 @@ impl ServiceStore {
             }
 
             // Remove the endpoint from the service.
-            if let Some(wls) = self.by_vip.get_mut(vip) {
-                wls.write().unwrap().endpoints.remove(addr);
+            if let Some(service) = self.by_vip.get(vip) {
+                // Clone the service and remove the endpoint.
+                let mut service = service.deref().clone();
+                service.endpoints.remove(addr);
+
+                // Update the service.
+                self.insert(service);
             }
         }
     }
 
     /// Adds the given service.
-    pub(super) fn insert(&mut self, mut service: Service) {
+    pub fn insert(&mut self, mut service: Service) {
         // First mutate the service and add all missing endpoints
-        for addr in service.addresses.as_slice() {
+        for vip in &service.vips {
             // Due to ordering issues, we may have gotten workloads with VIPs before we got the service
             // we should add those workloads to the vips map now
-            if let Some(wips_to_endpoints) = self.staged_vips.remove(addr) {
-                for (wip, ep) in wips_to_endpoints {
-                    service.endpoints.insert(wip.clone(), ep);
-                }
-            }
-
-            // If svc already exists, add old endpoints to new svc
-            if let Some(prev) = self.by_vip.get_mut(addr) {
-                let prev = prev.read().unwrap();
-                for (wip, ep) in prev.endpoints.iter() {
-                    service.endpoints.insert(wip.clone(), ep.clone());
+            if let Some(endpoints) = self.staged_vips.remove(vip) {
+                for (wip, ep) in endpoints {
+                    // Don't override if the new service has the endpoint.
```

Comment:
can we add a debug log line (or perhaps warning) in the event this happens? I'd think this is a bug not a race if we hit this scenario